bigtable/client/rpc_retry_policy.h

```diff
@@ -54,45 +53,72 @@ class RPCRetryPolicy {
+   */
+  virtual void setup(grpc::ClientContext& context) const = 0;
+
   /**
    * Handle an RPC failure.
    *
    * @return true if the RPC operation should be retried.
-   * @param delay how long to wait before retrying the operation.
    */
-  virtual bool on_failure(grpc::Status const& status,
-                          std::chrono::milliseconds* delay) = 0;
+  virtual bool on_failure(grpc::Status const& status) = 0;
+
+  /// Return true if the status code is retryable.
+  virtual bool can_retry(grpc::StatusCode code) const = 0;
 };
 
 /// Return an instance of the default RPCRetryPolicy.
 std::unique_ptr<RPCRetryPolicy> DefaultRPCRetryPolicy();
 
 /**
- * Implement a simple exponential backoff policy.
+ * Implement a simple "count errors and then stop" retry policy.
  */
-class ExponentialBackoffPolicy : public RPCRetryPolicy {
+class LimitedErrorCountRetryPolicy : public RPCRetryPolicy {
  public:
-  template <typename duration_t1, typename duration_t2>
-  ExponentialBackoffPolicy(int maximum_failures, duration_t1 initial_delay,
-                           duration_t2 maximum_delay)
-      : failure_count_(0),
-        maximum_failures_(maximum_failures),
-        current_delay_(std::chrono::duration_cast<std::chrono::microseconds>(
-            initial_delay)),
-        maximum_delay_(std::chrono::duration_cast<std::chrono::microseconds>(
-            maximum_delay)) {}
+  explicit LimitedErrorCountRetryPolicy(int maximum_failures)
+      : failure_count_(0), maximum_failures_(maximum_failures) {}
 
   std::unique_ptr<RPCRetryPolicy> clone() const override;
-  bool on_failure(grpc::Status const& status,
-                  std::chrono::milliseconds* delay) override;
+  void setup(grpc::ClientContext& context) const override;
+  bool on_failure(grpc::Status const& status) override;
+  bool can_retry(grpc::StatusCode code) const override;
 
  private:
   int failure_count_;
   int maximum_failures_;
-  std::chrono::microseconds current_delay_;
-  std::chrono::microseconds maximum_delay_;
 };
 
+/**
+ * Implement a simple "keep trying for this time" retry policy.
+ */
+class LimitedTimeRetryPolicy : public RPCRetryPolicy {
+ public:
+  template <typename duration_t>
+  explicit LimitedTimeRetryPolicy(duration_t maximum_duration)
+      : maximum_duration_(std::chrono::duration_cast<std::chrono::milliseconds>(
+            maximum_duration)),
+        deadline_(std::chrono::system_clock::now() + maximum_duration_) {}
+
+  std::unique_ptr<RPCRetryPolicy> clone() const override;
+  void setup(grpc::ClientContext& context) const override;
+  bool on_failure(grpc::Status const& status) override;
+  bool can_retry(grpc::StatusCode code) const override;
+
+ private:
+  std::chrono::milliseconds maximum_duration_;
+  std::chrono::system_clock::time_point deadline_;
+};
+
+/// The most common retryable codes, refactored because it is used in several
+/// places.
+constexpr bool IsIdempotentStatusCode(grpc::StatusCode code) {
```

Comment:
nit(s): Is there a reason we'd want to retry on `OK`? Also, I don't think anything about `DEADLINE_EXCEEDED`, for example, implies that retrying would be idempotent. `IsRetrableStatusCode` seems like a more accurate name?

Reply:
Done.